### PR TITLE
Make operator names consistent between export_opnames and the lite interpreter

### PR DESCRIPTION
**Pull request by iseeyuan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34674 Make operator names consistent between export_opnames and the lite interpreter**

Two changes to make sure the op_names dumped in export_opnames() are consistent to what are actually used in bytecode.
* Inline graph before dumping the operator names. 
* Use code of the graph (which is used in bytecode) instead of the nodes of graph.

Differential Revision: [D20610715](https://our.internmc.facebook.com/intern/diff/D20610715)

### CODE REVIEW

**Comment by ljk53** [comment]
LGTM - @bhosmer what do you think?

**Comment by bhosmer** [comment]
> LGTM - @bhosmer what do you think?

Looks fine to me, though I'm mostly navigating by other examples in the codebase that look like they're doing similar traversals (like [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/export_module.cpp#L42)). 

Only question that comes to mind is whether inlining like this is expensive, esp given that we're traversing a tree of modules. If so, could we recurse through `CALL` instructions instead? (If you're just using the prevailing idiom here please disregard - kind of just thinking out loud here.)

**Comment by iseeyuan** [comment]
> > LGTM - @bhosmer what do you think?
> 
> Looks fine to me, though I'm mostly navigating by other examples in the codebase that look like they're doing similar traversals (like [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/export_module.cpp#L42)).
> 
> Only question that comes to mind is whether inlining like this is expensive, esp given that we're traversing a tree of modules. If so, could we recurse through `CALL` instructions instead? (If you're just using the prevailing idiom here please disregard - kind of just thinking out loud here.)

The inlining are used for exporting only (either exporting the bytecode, or exporting the opnames list). It's done on server side once. So the performance would not be an issue for most models?

Not sure if we should go to instruction level to collect the op name list, but when there's a chance I may investigate the usage of CALL, to optimize the performance.

**Comment by bhosmer** [comment]
> The inlining are used for exporting only (either exporting the bytecode, or exporting the opnames list). It's done on server side once. So the performance would not be an issue for most models?
> 
> Not sure if we should go to instruction level to collect the op name list, but when there's a chance I may investigate the usage of CALL, to optimize the performance.

Ah actually I think it's definitely better if the bytecode and opnames exports worked the same way. Is there any way to actually reuse the same traversal for both? (Regardless, makes sense that inlining everything once at export time is probably fine.)

**Comment by iseeyuan** [comment]
> > The inlining are used for exporting only (either exporting the bytecode, or exporting the opnames list). It's done on server side once. So the performance would not be an issue for most models?
> > Not sure if we should go to instruction level to collect the op name list, but when there's a chance I may investigate the usage of CALL, to optimize the performance.
> 
> Ah actually I think it's definitely better if the bytecode and opnames exports worked the same way. Is there any way to actually reuse the same traversal for both? (Regardless, makes sense that inlining everything once at export time is probably fine.)

Good point. I have updated the code, trying to reuse as much as possible the existing logic.

File: torch/csrc/jit/serialization/export.cpp [hunk 1]
```diff
@@ -915,29 +916,33 @@ void check_onnx_proto(const std::string& proto_string) {
     onnx::checker::check_model(model);
 }
 
+void moduleMethodsTuple(
```
Comment (bhosmer) [inline torch/csrc/jit/serialization/export.cpp]: Is there a header file in common between `import.cpp` and `export_module.cpp` that we could put this declaration in?

File: torch/csrc/jit/serialization/export.cpp [hunk 2]
```diff
@@ -915,29 +916,33 @@ void check_onnx_proto(const std::string& proto_string) {
...
+    const Module& module,
+    std::vector<c10::IValue>& elements);
+IValue expect_field(IValue tup, const std::string& expected_name, size_t entry);
```
Comment (bhosmer) [inline torch/csrc/jit/serialization/export.cpp]: Same ask here, for a declaration in a common header.

File: torch/csrc/jit/serialization/export.cpp [hunk 3]
```diff
@@ -915,29 +916,33 @@ void check_onnx_proto(const std::string& proto_string) {
...
-        if (!opname.overload_name.empty()) {
-          namestr += "." + opname.overload_name;
-        }
-        opnames.emplace(namestr);
-      }
+void export_opnames(const script::Module& m, std::set<std::string>& opnames) {
+  std::vector<c10::IValue> elements;
+  moduleMethodsTuple(m, elements);
+  for (const auto& element : elements) {
+    auto table = element.toTuple()->elements()[1];
```
Comment (bhosmer) [inline torch/csrc/jit/serialization/export.cpp]: I worry about the use of synchronized raw numbers to spider into these tuples (i.e., both here and in `parseMethods`, maybe elsewhere?). 

Maybe it's overkill to factor this logic into a common utility, but I think it'd be good to at least make these tuple indexes into a common `#define`s. WDYT?

**Review by bhosmer — COMMENTED** [review]
Excellent! Feels good to be using the common utilities. A couple fairly minor things inline and then I think it's good to go. Thanks!

File: torch/csrc/jit/serialization/export.cpp (hunk 1 above)
Comment (iseeyuan) [reply torch/csrc/jit/serialization/export.cpp]: The common header is fine, but may be unnecessarily included in other files. Any disadvantage of declaration?

Comment (bhosmer) [reply torch/csrc/jit/serialization/export.cpp]: Only that someone changing the signature of the implementation would need to know to update this declaration - I'd probably assume that if a function were used publicly, there would be a declaration in a header somewhere. But doing it this way might be more of a convention in this codebase. Are you following an existing idiom? I'm fine with it if so.

**Review by bhosmer — APPROVED** [review]
This looks great! The external decls are fine by me if it's standard style to do this - per inline comment I'd assumed that external declarations were always done in headers but I could definitely be wrong.

**Comment by iseeyuan** [comment]
> This looks great! The external decls are fine by me if it's standard style to do this - per inline comment I'd assumed that external declarations were always done in headers but I could definitely be wrong.

I think you're right. Didn't realize that there is such a convention. Just moved them to the header file.

**Pull request closed without merging**